File: src/DomainF/Tournaments/Rules.fs
module Orleans.Tournament.Domain.Tournaments.Rules
open Orleans.Tournament.Domain.Helpers
open Orleans.Tournament.Domain.Tournaments

let TournamentExists (state : TournamentState) =
    match state.Created with
    | true -> Ok ()
    | false -> Error BusinessErrors.TournamentDoesNotExist

let TournamentDoesNotExists (state : TournamentState) =
    match state.Created with
    | false -> Ok ()
    | true -> Error BusinessErrors.TournamentAlreadyExist

let TeamExistsForward exists =
    match exists with
    | true -> Ok ()
    | false -> Error BusinessErrors.TeamDoesNotExist

let TeamIsNotAdded (state : TournamentState) teamId =
    match state.Teams |> List.tryFind (fun e -> e = teamId) with
    | Some _ -> Error BusinessErrors.TeamIsAlreadyAdded
    | None -> Ok ()

let LessThanEightTeams (state : TournamentState) =
    match state.Teams.Length with
    | c when c < 8 -> Ok ()
    | _ -> Error BusinessErrors.TournamentHasMoreThanEightTeams

let EightTeamsToStartTournament (state : TournamentState) =
    match state.Teams.Length = 8 with
    | true -> Ok ()
    | false -> Error BusinessErrors.CantStartTournamentWithLessThanEightTeams

let TournamentIsStarted (state : TournamentState) =
    match state.Fixture with
    | Some _ -> Ok ()
    | None -> Error BusinessErrors.TournamentIsNotStarted

let TournamentIsNotStarted (state : TournamentState) =
    match state.Fixture with
    | None -> Ok ()
    | Some _ -> Error BusinessErrors.TournamentAlreadyStarted

let MatchExistsAndIsNotPlayed (state : TournamentState) matchInfo =
    let currentPhase fixture =
        match fixture.Finals with
        | Some f -> f
        | None ->
            match fixture.SemiFinals with
            | Some sf -> sf
            | None -> fixture.QuarterFinals

    match state.Fixture with
    | Some f ->
        let c = currentPhase f
        let m = c.Matches |> List.tryFind (fun e -> e.IsMatchInfo matchInfo)
        match m with
        | Some m ->
            match m.MatchSummary with
            | None -> Ok ()
            | Some _ -> Error BusinessErrors.MatchAlreadyPlayed
        | None -> Error BusinessErrors.MatchDoesNotExist
    | None -> Error BusinessErrors.TournamentIsNotStarted

let MatchIsNotDraw (matchSummary : MatchSummary) =
    if matchSummary.LocalGoals = matchSummary.AwayGoals then true
    else false

File: src/Domain/Tournaments/Rules.cs
namespace Tournament.Domain.Tournaments;

public partial class TournamentState
{
    public Results TournamentExists()
        => Created
            ? Results.Unit
            : Results.TournamentDoesNotExist;

    public Results TournamentDoesNotExists()
        => !Created
            ? Results.Unit
            : Results.TournamentAlreadyExist;

    public Results TeamIsNotAdded(Guid teamId)
        => !Teams.Contains(teamId)
            ? Results.Unit
            : Results.TeamIsAlreadyAdded;

    public Results LessThanEightTeams()
        => Teams.Count < 8
            ? Results.Unit
            : Results.TournamentHasMoreThanEightTeams;

    public Results EightTeamsToStartTournament()
        => Teams.Count == 8
            ? Results.Unit
            : Results.TournamentCantStartWithLessThanEightTeams;

    public Results TournamentStarted()
        => Fixture.Quarter.IsEmpty == false
            ? Results.Unit
            : Results.TournamentDidNotStart;

    public Results TournamentDidNotStart()
        => Fixture.Quarter.IsEmpty
            ? Results.Unit
            : Results.TournamentAlreadyStarted;

    public Results MatchExistsAndIsNotPlayed(Match match)
    {
        if (Fixture.Quarter.IsEmpty)
            return Results.TournamentDidNotStart;

        // Get current phase
        var currentPhase = Fixture.Quarter;

        if (!Fixture.Final.IsEmpty)
            currentPhase = Fixture.Final;
        else if (!Fixture.Semi.IsEmpty)
            currentPhase = Fixture.Semi;

        var currentMatch = currentPhase.Matches.SingleOrDefault(e =>
            e.LocalTeamId == match.LocalTeamId &&
            e.AwayTeamId == match.AwayTeamId);

        if (currentMatch is null)
            return Results.MatchDoesNotExist;

        if (currentMatch.MatchResult.Played)
            return Results.MatchAlreadyPlayed;

        return Results.Unit;
    }

    public Results MatchIsNotDraw(MatchResult matchResult)
        => matchResult.LocalGoals == matchResult.AwayGoals
            ? Results.DrawResultIsNotAllowed
            : Results.Unit;
}
